File: pybot/test_dispersion.py
import psyco
import random
import time

psyco.full()
# apply simple dispersion function to grid
rows = 200
cols = 200
grid = [[random.random() for i in xrange(cols)] for j in xrange(rows)]
def disperse(grid):
	new_grid = [[0 for c in xrange(cols)] for r in xrange(rows)]
	for row in xrange(rows):
		for col in xrange(cols):
			if row > 0 and row < (rows - 1) and col > 0 and col < (cols - 1):
				center_energy = grid[row][col]
				adjsum = 0
				adjsum += grid[row - 1][col] - center_energy
				adjsum += grid[row + 1][col] - center_energy
				adjsum += grid[row][col + 1] - center_energy
				adjsum += grid[row][col - 1] - center_energy
				adjsum /= 4.0
				new_val = 0.5 * (center_energy + adjsum)
				new_grid[row][col] = new_val
	return new_grid


start = time.time()
for i in xrange(10):
	#for r in xrange(rows):
	#	for c in xrange(cols):
	#		x = grid[r][c]
	grid = disperse(grid)
print "elapsed: {0}".format(time.time() - start)

File: test_dispersion.c
int main() {
int ROWS = 150;
int COLS = 150;

float grid[ROWS][COLS];
int blah;

int iter = 0;
int r = 0;
int c = 0;
for (iter = 0; iter < 20; iter++) {
	for (r = 0; r < ROWS; r++) {
		for (c = 0; c < COLS; c++) {
			float center_energy = grid[r][c];
			float nigadj = 0;
			if (r > 0 && r < (ROWS - 1) && c > 0 && c < (COLS - 1)) {
			nigadj += grid[r - 1][c] - center_energy;
			nigadj += grid[r + 1][c] - center_energy;
			nigadj += grid[r][c + 1] - center_energy;
			nigadj += grid[r][c - 1] - center_energy;
			nigadj /= 4.0;
			float new_val = 0.5 * (center_energy + nigadj);
			grid[r][c] = new_val;
			}
		}
	}
}

return 0;
}
